documentation/prometheus-design-principles.md

```diff
@@ -0,0 +1,164 @@
+# Prometheus design principles
+
+What is prometheus for?  There seem to be a number of design
+principles behind prometheus that aren't explicitly documented on the
+website.  However, prometheus maintainers sometimes refer to them in
+mailing lists or github issues.  This page tries to reverse engineer
+them.
+
+## Terminology
+
+The Prometheus community uses certain words as jargon terms.  These
+are some:
+
+  * failure domain: a group of things that can all fail at once (a
+    data centre, an AWS region or AZ)
+  * instance: an endpoint that you can scrape
+  * job: a collection of *instances* with the same purpose, replicated
+    for scalability or reliability or some other purpose
+
+## What Prometheus does
+
+Prometheus is a tool for collecting time-series data.
+
+  * prometheus is not designed for long-term archival
+      * though in practice, prometheis with a history of a year or
+        more can and do exist
+      * it's also not designed to be sure you've captured every data
+        point
+
+## What Prometheus doesn't do
+
+[Julius Volz listed some things Prometheus doesn't do](https://youtu.be/QgJbxCWRZ1s?t=9m28s):
+
+  * raw log / event collection (use a logging system)
+  * request tracing (use zipkin or similar)
+  * "magic" anomaly detection
+  * durable long-term storage (use an external system for this,
+    decouple it from your operational monitoring)
+  * automatic horizontal scaling
+  * user / auth management
+
+## Pull model
+
+Prometheus is a pull-based system.
+[Julius Volz listed some benefits of pull](https://youtu.be/QgJbxCWRZ1s?t=17m55s):
+
+  * automatic upness monitoring - if you fail to scrape metrics, you
+    can immediately say something is wrong
+  * "horizontal monitoring" - Team A can include some metrics from
+    Team B on their operational dashboard by selectively scraping.
+    You can be flexible with what you pull from where
+  * run a copy of monitoring on your laptop, scraping the same endpoints
+      * experiment with different alerting rules
+  * simpler HA - just deploy a copy (alertmanager deduplicates alerts)
+  * less configuration - instances don't need to know where monitoring
```

Comment:
+ The monitoring system is in control of when it ingests new metrics by setting a scraping interval. This avoids issues where a misbehaving client emits so many metrics they negatively impact other users of the monitoring system.